File: javascript/json-transform/src/functions/TransformerFunctionRepeat.ts
import { createAsyncSequence } from "@wortise/sequency";
import TransformerFunction from "./common/TransformerFunction";
import { ArgType } from "./common/ArgType";
import FunctionContext from "./common/FunctionContext";
import JsonElementStreamer from "../JsonElementStreamer";

class TransformerFunctionRepeat extends TransformerFunction {
  constructor() {
    super({
      arguments: {
        count: { type: ArgType.Integer, position: 0 },
      },
    });
  }

  override async apply(context: FunctionContext): Promise<any> {
    const value = await context.getJsonElement(null);
    if (value === null) {
      return null;
    }
    const count = await context.getInteger("count");
    if (count == null || count < 0) {
      return null;
    }

    return JsonElementStreamer.fromTransformedStream(
      context,
      createAsyncSequence({
        next: () => Promise.resolve({ value }),
      }).take(count),
    );
  }
}

export default TransformerFunctionRepeat;

File: java/json-transform/src/main/java/co/nlighten/jsontransform/functions/TransformerFunctionRepeat.java
package co.nlighten.jsontransform.functions;

import co.nlighten.jsontransform.JsonElementStreamer;
import co.nlighten.jsontransform.functions.common.*;

import java.util.Map;
import java.util.stream.Stream;

public class TransformerFunctionRepeat extends TransformerFunction {
    public TransformerFunctionRepeat() {
        super(FunctionDescription.of(
            Map.of(
                "count", ArgumentType.of(ArgType.Integer).position(0)
            )
        ));
    }

    @Override
    public Object apply(FunctionContext context) {
        var value = context.getJsonElement(null);
        if (value == null) {
            return null;
        }
        var count = context.getInteger("count");
        if (count == null || count < 0) {
            return null;
        }

        return JsonElementStreamer.fromTransformedStream(context,
            Stream.generate(() -> value).limit(count)
        );
    }
}
